src/propulsion/controller.cpp

```diff
@@ -0,0 +1,792 @@
+{
+  return true;
+}
+
+bool Controller::sendControllerMessage(ControllerMessage message_template)
+{
+  for (int i = 0; i < message_template.len; i++) {
+    sdo_message_.data[i] = message_template.message_data[i];
+  }
+
+  sendSdoMessage(sdo_message_);
+
+  log_.DBG1("MOTOR", message_template.logger_output, node_id_);
+
+  if (critical_failure_) {
+    return true;
+  }
+  return false;
+}
+
+void Controller::registerController()
+{
+  sender.registerController();
+}
+
+void Controller::configure()
+{
+  log_.INFO("MOTOR", "Controller %d: Configuring...", node_id_);
+  for (int i = 0; i < 16; i++) {
+    if (sendControllerMessage(configMsgs_[i])) return;
+  }
+  log_.INFO("MOTOR", "Controller %d: Configured.", node_id_);
+}
+
+void Controller::enterOperational()
+{
+  // TODO(Iain): Check that this is still valid:
+
+  // Send NMT Operational message to transition from state 0 (Not ready to switch on)
+  // to state 1 (Switch on disabled)
+  nmt_message_.data[0] = kNmtOperational;
+  nmt_message_.data[1] = node_id_;
+
+  log_.INFO("MOTOR", "Controller %d: Sending NMT Operational command", node_id_);
+  sender.sendMessage(nmt_message_);
+  // leave time for the controller to enter NMT Operational
+  Thread::sleep(100);
+
+  // enables velocity mode
+  if (sendControllerMessage(enterOpMsgs_[0])) return;
+
+  // set the velocity to zero
+  sendTargetVelocity(0);
+
+  // apply break
+  if (sendControllerMessage(enterOpMsgs_[1])) return;
+
+  // send shutdown message to transition to Ready to Switch On state
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = enterOpMsgs_[2].message_data[i];
+  }
+  log_.DBG1("MOTOR", enterOpMsgs_[2].logger_output, node_id_);
+  requestStateTransition(sdo_message_, kReadyToSwitchOn);
+
+  // Send enter operational message to transition to the Operation Enabled state
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = enterOpMsgs_[3].message_data[i];
+  }
+  log_.DBG1("MOTOR", enterOpMsgs_[3].logger_output, node_id_);
+  requestStateTransition(sdo_message_, kOperationEnabled);
+}
+
+void Controller::enterPreOperational()
+{
+  checkState();
+  if (state_ != kReadyToSwitchOn) {
+    // send shutdown command
+    if (sendControllerMessage(enterPreOpMsg_[0])) return;
+  }
+}
+
+void Controller::checkState()
+{
+  // Check Statusword in object dictionary
+  if (sendControllerMessage(checkStateMsg_[0])) return;
+}
+
+void Controller::sendTargetVelocity(int32_t target_velocity)
+{
+  // Send 32 bit integer in Little Edian bytes
+  for (int i = 0; i < 8; i++) {
+    sdo_message_.data[i] = sendTargetVelMsg[0].message_data[i];
+  }
+  sdo_message_.data[4] = target_velocity & 0xFF;
+  sdo_message_.data[5] = (target_velocity >> 8) & 0xFF;
+  sdo_message_.data[6] = (target_velocity >> 16) & 0xFF;
+  sdo_message_.data[7] = (target_velocity >> 24) & 0xFF;
+
+  log_.DBG2("MOTOR", sendTargetVelMsg[0].logger_output, node_id_, target_velocity);
+  sender.sendMessage(sdo_message_);
```

Comment:
we don't wait for a return value for this can transmission. Since sendSdoMessage will wait for a response, it would just slow the acceleration process down